### PR TITLE
HPCC-27965 Remove extra log message from initWULogReader()

### PR DESCRIPTION
**Pull request by wangkx**
Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: esp/services/ws_workunits/ws_workunitsHelpers.cpp
Comment (jakesmith) [inline esp/services/ws_workunits/ws_workunitsHelpers.cpp:2131]: This means that if the configuration changes (e.g. on a helm update) it may log that it failed, but subsequently succeed (silently) or the reverse..

I think it would be better to either remove the tracing altogether or add the logic inside the singleton method in jlog, but also spot when the config changes and update via a ConfigUpdateHook (example use case in eclccserver.cpp, line 1099)

**Review by rpastrana — COMMENTED** [review]
@wangkx couple of comments

File: esp/services/ws_workunits/ws_workunitsHelpers.cpp
Comment (jakesmith) [reply esp/services/ws_workunits/ws_workunitsHelpers.cpp:2130]: Yes, that was what I saying in my comment too : https://github.com/hpcc-systems/HPCC-Platform/pull/16343#discussion_r919168991

File: esp/services/ws_workunits/ws_workunitsHelpers.cpp
Comment (rpastrana) [reply esp/services/ws_workunits/ws_workunitsHelpers.cpp:2131]: @jakesmith ConfigUpdateHook looks to be very useful, i think I've mentioned this before but these types of features need to be publicized loudly to the entire development team!

Comment (wangkx) [reply esp/services/ws_workunits/ws_workunitsHelpers.cpp:2131]: It seems that the reloadConfigHook.installOnce(std::bind(&CWsWorkunitsEx::initWULogReader, this), false); does not run the  initWULogReader() itself.

Comment (rpastrana) [reply esp/services/ws_workunits/ws_workunitsHelpers.cpp:2131]: @wangkx I see no issue keeping the logging in the init method if only called at load/reconfig time

**Review by rpastrana — APPROVED** [review]
@wangkx looks reasonable. 
BTW, I'll handle the related issue discussed offline as part of this change: https://track.hpccsystems.com/browse/HPCC-27983

File: esp/services/ws_workunits/ws_workunitsHelpers.cpp
Comment (jakesmith) [inline esp/services/ws_workunits/ws_workunitsHelpers.cpp:209]: we should avoid not compiling container centric code, where possible.
Keeping it general and compiling in all builds, helps spot issues/has more coverage.

But also, in this case remote logs is not necessarily just a containerized feature, it could equally apply to bare-metal (if given configuration support etc.), so we should not exclude it upfront.

File: esp/services/ws_workunits/ws_workunitsHelpers.cpp
Comment (jakesmith) [inline esp/services/ws_workunits/ws_workunitsHelpers.cpp:4069]: see previous comment, better not to exclude unless there's a particular reason

File: esp/services/ws_workunits/ws_workunitsHelpers.hpp
Comment (jakesmith) [inline esp/services/ws_workunits/ws_workunitsHelpers.hpp:186]: see prev. comments re. not excluding

**Review by jakesmith — COMMENTED** [review]
@wangkx - please see comments.

File: esp/services/ws_workunits/ws_workunitsAuditLogs.cpp
Comment (wangkx) [reply esp/services/ws_workunits/ws_workunitsAuditLogs.cpp:1448]: @jakesmith @rpastrana If I understand the comment correctly, I should roll back all of the changes in https://github.com/hpcc-systems/HPCC-Platform/pull/16343/commits/64e9ea37773bd6b9946802b739544e57ee07c4bd. And remove the LOG line and the OWARNLOG line from initWULogReader() for now. When we revise the Singleton (https://track.hpccsystems.com/browse/HPCC-27983), we may handle the config changes, add the tracing, return a pointer, etc. within the Singleton.

Comment (jakesmith) [reply esp/services/ws_workunits/ws_workunitsAuditLogs.cpp:1448]: Requirements for HPCC-27983 go beyond the requirements for handling a single initialization and ptr return from queryRemoteLogAccessor(). HPCC-27983 is to auto-detect configuration changes and to adjust (if necessary) the underlying object that queryRemoteLogAccessor() returns.

Comment (jakesmith) [reply esp/services/ws_workunits/ws_workunitsAuditLogs.cpp:1448]: I've pushed a pending change to allow the Singleton helper to cope with being initialized (once) to null: https://track.hpccsystems.com/browse/HPCC-27988

Comment (jakesmith) [reply esp/services/ws_workunits/ws_workunitsAuditLogs.cpp:1448]: > And remove the LOG line and the OWARNLOG line from initWULogReader() for now.

I would be happy with that in the short term if @rpastrana - not sure the logging is _that_ useful that we'd miss it, but it's also fine/better to be issued once only (as long as it doesn't complicate all the code that uses it)

Once HPCC-27988 is merged, I'd suggest queryRemoteLogAccessor is changed to return a ptr (and it is valid for it return null), and it can log 1st time initialization and/or exceptions before returning.

That change can be made ahead of work for HPCC-27983 - which is going to be a bit more work.

Comment (jakesmith) [reply esp/services/ws_workunits/ws_workunitsAuditLogs.cpp:1448]: then let's wait for HPCC-27988 to be merged, and have the factory method (which will then only be called once) issue the logging so it's universal.

Comment (jakesmith) [reply esp/services/ws_workunits/ws_workunitsAuditLogs.cpp:1448]: @wangkx https://github.com/hpcc-systems/HPCC-Platform/pull/16368 (HPCC-27988) is now merged.

**Comment by wangkx** [comment]
@jakesmith I made the changes. Please take a look.

**Review by jakesmith — DISMISSED** [review]
@wangkx - the idea was after HPCC-27988 was merged (related to the conversation at https://github.com/hpcc-systems/HPCC-Platform/pull/16343#discussion_r922096792),
for the logInitWULogReader to be redundant,
instead for the initialization logging, or failure logging, to go directly into the factor method (the lambda func. in jlog.cpp - line 3246).

Then callers can cheaply call queryRemoteLogAccessor() and if necessary check if it returns nullptr. No separate storage of a secondary pointer to a IRemoteLogAccess

**Comment by wangkx** [comment]
@jakesmith please review

File: esp/services/ws_logaccess/WsLogAccessService.cpp
Comment (jakesmith) [inline esp/services/ws_logaccess/WsLogAccessService.cpp:14]: future: fine as it is now, but when or if an auto-updating mechanism is added (HPCC-27983), this will method will likely want to turn into getRemoteLogAccessor, so that it links the object, and store it in a temporary Owned, for use below. That way the update can change the object, whilst the existing object is still valid.

File: system/jlib/jlog.cpp
Comment (jakesmith) [inline system/jlib/jlog.cpp:3246]: I think, as before, should keep a 'loading..' type message (that previously was in initWULogReader and Cws_logaccessEx::init), at the start of this function.
  PROGLOG("Loading remote log access plug-in.");

**Review by jakesmith — COMMENTED** [review]
@wangkx - looks much cleaner, 1 minor comment

**Review by jakesmith — APPROVED** [review]
@wangkx - looks good, please squash

**Comment by wangkx** [comment]
@richardkchapman this PR is ready to be merged.

**Pull request merged**
